lib/katello/tasks/clean_backend_objects.rake

```diff
@@ -1,42 +1,77 @@
 namespace :katello do
   desc "Cleans backend objects (hosts) that are missing in one or more backend systems.  Run with COMMIT=true to commit changes."
   task :clean_backend_objects => ["environment", "check_ping"] do
-    def cleanup_hosts
-      Host.includes(:content_facet, :subscription_facet).find_each do |host|
-        if test_candlepin(host) || test_pulp(host)
-          print "Host #{host.id} #{host.name} #{host.subscription_facet.try(:uuid)} is partially missing.  Un-registering\n"
-          execute("Failed to delete host") { ForemanTasks.sync_task(::Actions::Katello::Host::Unregister, host) }
+    class BackendCleaner
+      def initialize
+        @candlepin_uuids = []
+        @pulp_uuids = []
+        @katello_candlepin_uuids = []
+        @katello_pulp_uuids = []
+      end
+
+      def populate!
+        @candlepin_uuids = Katello::Resources::Candlepin::Consumer.all_uuids
+        @katello_candlepin_uuids = Katello::Host::SubscriptionFacet.pluck(:uuid).compact
+
+        @pulp_uuids = ::Katello.pulp_server.extensions.consumer.retrieve(nil).map { |consumer| consumer['id'] }
+        @katello_pulp_uuids = Katello::Host::ContentFacet.pluck(:uuid).compact
+      end
+
+      def hosts_with_no_subscriptions
+        ::Host.where(:id => Katello::Host::SubscriptionFacet.where(:uuid => @katello_candlepin_uuids - @candlepin_uuids).select(:host_id))
```

Comment:
I like these kind of things on 2 lines, but I'll leave that up to you if you want to change (not a blocker)
```ruby
sub_facet_ids = Katello::Host::SubscriptionFacet.where(:uuid => @katello_candlepin_uuids - @candlepin_uuids).select(:host_id)
::Host.where(:id => sub_facet_ids)
```